internal/experiment/webconnectivityqa/dnshijacking.go

```diff
@@ -33,11 +33,11 @@ func dnsHijackingToProxyWithHTTPURL() *TestCase {
 			StatusCodeMatch: true,
 			HeadersMatch:    true,
 			TitleMatch:      true,
-			XStatus:         2, // StatusSuccessCleartext
-			XDNSFlags:       0,
-			XBlockingFlags:  32, // analysisFlagSuccess
-			Accessible:      true,
-			Blocking:        false,
+			XStatus:         2,     // StatusSuccessCleartext
+			XDNSFlags:       4,     // AnalysisDNSUnexpectedAddrs
+			XBlockingFlags:  33,    // AnalysisFlagDNSBlocking | analysisFlagSuccess
+			Accessible:      false, // FIXME: this is probably incorrect!
+			Blocking:        "dns",
```

Comment:
This case seems tricky. We can see the result oscillating. In some cases, we get the expectations that are listed above, which changed since `master`, and which we'll call case `A`. In some other cases, we see the following error, which we'll call case `B`:

```
qa_test.go:20: test keys mismatch:   &webconnectivityqa.testKeys{
          	... // 1 ignored field
          	DNSExperimentFailure: nil,
        - 	DNSConsistency:       string("inconsistent"),
        + 	DNSConsistency:       string("consistent"),
          	ControlFailure:       nil,
          	... // 3 ignored and 4 identical fields
        - 	XDNSFlags:      4,
        + 	XDNSFlags:      0,
        - 	XBlockingFlags: 33,
        + 	XBlockingFlags: 32,
          	XNullNullFlags: 0,
        - 	Accessible:     bool(false),
        + 	Accessible:     bool(true),
        - 	Blocking:       string("dns"),
        + 	Blocking:       bool(false),
          }
```

In the former case, the analysis is this:

```JSON
{
  "DNSExperimentFailure": null,
  "DNSTransactionsWithBogons": {},
  "DNSTransactionsWithUnexpectedFailures": {},
  "DNSPossiblyInvalidAddrs": {
    "130.192.182.17": true
  },
  "HTTPDiffBodyProportionFactor": 1,
  "HTTPDiffStatusCodeMatch": true,
  "HTTPDiffTitleDifferentLongWords": {},
  "HTTPDiffUncommonHeadersIntersection": {
    "content-length": true
  },
  "HTTPFinalResponses": {
    "4": true
  },
  "HTTPFinalResponsesWithTLS": {},
  "TCPTransactionsWithUnexpectedTCPConnectFailures": {},
  "TCPTransactionsWithUnexpectedTLSHandshakeFailures": {},
  "TCPTransactionsWithUnexpectedHTTPFailures": {},
  "TCPTransactionsWithUnexplainedUnexpectedFailures": {}
}
```

In the latter case, the analysis is:

```JSON
{
  "DNSExperimentFailure": null,
  "DNSTransactionsWithBogons": {},
  "DNSTransactionsWithUnexpectedFailures": {},
  "DNSPossiblyInvalidAddrs": {},
  "HTTPDiffBodyProportionFactor": 1,
  "HTTPDiffStatusCodeMatch": true,
  "HTTPDiffTitleDifferentLongWords": {},
  "HTTPDiffUncommonHeadersIntersection": {
    "content-length": true
  },
  "HTTPFinalResponses": {
    "4": true
  },
  "HTTPFinalResponsesWithTLS": {},
  "TCPTransactionsWithUnexpectedTCPConnectFailures": {},
  "TCPTransactionsWithUnexpectedTLSHandshakeFailures": {},
  "TCPTransactionsWithUnexpectedHTTPFailures": {},
  "TCPTransactionsWithUnexplainedUnexpectedFailures": {}
}
```

The difference between the two cases is:

```diff
--- A	2023-11-27 14:01:02
+++ B	2023-11-27 14:00:48
@@ -2,9 +2,7 @@
   "DNSExperimentFailure": null,
   "DNSTransactionsWithBogons": {},
   "DNSTransactionsWithUnexpectedFailures": {},
-  "DNSPossiblyInvalidAddrs": {
-    "130.192.182.17": true
-  },
+  "DNSPossiblyInvalidAddrs": {},
   "HTTPDiffBodyProportionFactor": 1,
   "HTTPDiffStatusCodeMatch": true,
   "HTTPDiffTitleDifferentLongWords": {},
```

The difference between the observations is the following:

```diff
@@ -40,12 +40,12 @@
       "ControlHTTPResponseHeadersKeys": null,
       "ControlHTTPResponseTitle": null
     },
-    "2": {
+    "3": {
       "DNSTransactionIDs": [
-        2
+        3
       ],
       "DNSDomain": "www.example.com",
-      "DNSLookupFailure": "dns_no_answer",
+      "DNSLookupFailure": "dns_nxdomain_error",
       "DNSQueryType": "AAAA",
       "DNSEngine": "doh",
       "IPAddress": null,
@@ -84,7 +84,7 @@
   "KnownTCPEndpoints": {
     "4": {
       "DNSTransactionIDs": [
-        3,
+        2,
         1
       ],
       "DNSDomain": "www.example.com",
@@ -134,23 +134,24 @@
     },
     "5": {
       "DNSTransactionIDs": [
-        2
+        2,
+        1
       ],
       "DNSDomain": "www.example.com",
       "DNSLookupFailure": "",
       "DNSQueryType": null,
       "DNSEngine": null,
-      "IPAddress": "93.184.216.34",
-      "IPAddressASN": 15133,
-      "IPAddressOrg": "Edgecast Inc.",
+      "IPAddress": "130.192.182.17",
+      "IPAddressASN": 137,
+      "IPAddressOrg": "Consortium GARR",
       "IPAddressBogon": false,
       "EndpointTransactionID": 5,
       "EndpointProto": "tcp",
-      "EndpointPort": "80",
-      "EndpointAddress": "93.184.216.34:80",
+      "EndpointPort": "443",
+      "EndpointAddress": "130.192.182.17:443",
       "TCPConnectFailure": "",
-      "TLSHandshakeFailure": null,
-      "TLSServerName": null,
+      "TLSHandshakeFailure": "",
+      "TLSServerName": "www.example.com",
       "HTTPRequestURL": null,
       "HTTPFailure": null,
       "HTTPResponseStatusCode": null,
@@ -163,9 +164,9 @@
       "ControlDNSDomain": null,
       "ControlDNSLookupFailure": null,
       "ControlTCPConnectFailure": "",
-      "MatchWithControlIPAddress": true,
-      "MatchWithControlIPAddressASN": true,
-      "ControlTLSHandshakeFailure": null,
+      "MatchWithControlIPAddress": false,
+      "MatchWithControlIPAddressASN": false,
+      "ControlTLSHandshakeFailure": "",
       "ControlHTTPFailure": "",
       "ControlHTTPResponseStatusCode": 200,
       "ControlHTTPResponseBodyLength": 1533,
@@ -178,25 +179,22 @@
       "ControlHTTPResponseTitle": "Default Web Page"
     },
     "6": {
-      "DNSTransactionIDs": [
-        3,
-        1
-      ],
-      "DNSDomain": "www.example.com",
-      "DNSLookupFailure": "",
+      "DNSTransactionIDs": [],
+      "DNSDomain": null,
+      "DNSLookupFailure": null,
       "DNSQueryType": null,
       "DNSEngine": null,
-      "IPAddress": "130.192.182.17",
-      "IPAddressASN": 137,
-      "IPAddressOrg": "Consortium GARR",
+      "IPAddress": "93.184.216.34",
+      "IPAddressASN": 15133,
+      "IPAddressOrg": "Edgecast Inc.",
       "IPAddressBogon": false,
       "EndpointTransactionID": 6,
       "EndpointProto": "tcp",
-      "EndpointPort": "443",
-      "EndpointAddress": "130.192.182.17:443",
+      "EndpointPort": "80",
+      "EndpointAddress": "93.184.216.34:80",
       "TCPConnectFailure": "",
-      "TLSHandshakeFailure": "",
-      "TLSServerName": "www.example.com",
+      "TLSHandshakeFailure": null,
+      "TLSServerName": null,
       "HTTPRequestURL": null,
       "HTTPFailure": null,
       "HTTPResponseStatusCode": null,
@@ -209,9 +207,9 @@
       "ControlDNSDomain": null,
       "ControlDNSLookupFailure": null,
       "ControlTCPConnectFailure": "",
-      "MatchWithControlIPAddress": false,
-      "MatchWithControlIPAddressASN": false,
-      "ControlTLSHandshakeFailure": "",
+      "MatchWithControlIPAddress": null,
+      "MatchWithControlIPAddressASN": null,
+      "ControlTLSHandshakeFailure": null,
       "ControlHTTPFailure": "",
       "ControlHTTPResponseStatusCode": 200,
       "ControlHTTPResponseBodyLength": 1533,
@@ -224,11 +222,9 @@
       "ControlHTTPResponseTitle": "Default Web Page"
     },
     "7": {
-      "DNSTransactionIDs": [
-        2
-      ],
-      "DNSDomain": "www.example.com",
-      "DNSLookupFailure": "",
+      "DNSTransactionIDs": [],
+      "DNSDomain": null,
+      "DNSLookupFailure": null,
       "DNSQueryType": null,
       "DNSEngine": null,
       "IPAddress": "93.184.216.34",
@@ -254,8 +250,8 @@
       "ControlDNSDomain": null,
       "ControlDNSLookupFailure": null,
       "ControlTCPConnectFailure": "",
-      "MatchWithControlIPAddress": true,
-      "MatchWithControlIPAddressASN": true,
+      "MatchWithControlIPAddress": null,
+      "MatchWithControlIPAddressASN": null,
       "ControlTLSHandshakeFailure": "",
       "ControlHTTPFailure": "",
       "ControlHTTPResponseStatusCode": 200,
```

Reply:
Interestingly, both runs seem to produce the same observations (as far as I can tell). So, either I am missing something or the algorithm producing the results must contain something subtle that depends on ordering.

Reply:
Fixed by https://github.com/ooni/probe-cli/commit/8d25a65d8eb13f4b457e3b29b8dd38f040ec4f74